File: library/src/main/java/com/pengrad/telegrambot/request/CreateInvoiceLink.java
package com.pengrad.telegrambot.request;

import com.pengrad.telegrambot.model.request.LabeledPrice;
import com.pengrad.telegrambot.response.StringResponse;

/**
 * Mirco Ianese
 * 21 Jun 2022
 */
public class CreateInvoiceLink extends BaseRequest<CreateInvoiceLink, StringResponse> {

    public CreateInvoiceLink(String title, String description, String payload, String providerToken,
                             String currency, LabeledPrice... prices) {
        super(StringResponse.class);
        add("title", title).add("description", description).add("payload", payload)
                .add("provider_token", providerToken).add("currency", currency).add("prices", prices);
    }


    /**
     * @param maxTipAmount The maximum accepted amount for tips in the smallest units of the currency
     */
    public CreateInvoiceLink maxTipAmount(int maxTipAmount) {
        return add("max_tip_amount", maxTipAmount);
    }

    /**
     * @param suggestedTipAmounts An array of suggested amounts of tip in the smallest units of the currency. At most 4 suggested tip amounts can be specified. The suggested tip amounts must be positive, passed in a strictly increased order and must not exceed max_tip_amount.
     */
    public CreateInvoiceLink suggestedTipAmounts(Integer[] suggestedTipAmounts) {
        return add("suggested_tip_amounts", suggestedTipAmounts);
    }

    public CreateInvoiceLink providerData(String providerData) {
        return add("provider_data", providerData);
    }

    public CreateInvoiceLink photoUrl(String photoUrl) {
        return add("photo_url", photoUrl);
    }

    public CreateInvoiceLink photoSize(Integer photoSize) {
        return add("photo_size", photoSize);
    }

    public CreateInvoiceLink photoWidth(Integer photoWidth) {
        return add("photo_width", photoWidth);
    }

    public CreateInvoiceLink photoHeight(Integer photoHeight) {
        return add("photo_height", photoHeight);
    }

    public CreateInvoiceLink needName(boolean needName) {
        return add("need_name", needName);
    }

    public CreateInvoiceLink needPhoneNumber(boolean needPhoneNumber) {
        return add("need_phone_number", needPhoneNumber);
    }

    public CreateInvoiceLink needEmail(boolean needEmail) {
        return add("need_email", needEmail);
    }

    public CreateInvoiceLink needShippingAddress(boolean needShippingAddress) {
        return add("need_shipping_address", needShippingAddress);
    }

    public CreateInvoiceLink sendEmailToProvider(boolean sendEmailToProvider) {
        return add("send_email_to_provider", sendEmailToProvider);
    }

    public CreateInvoiceLink sendPhoneNumberToProvider(boolean sendPhoneNumberToProvider) {
        return add("send_phone_number_to_provider", sendPhoneNumberToProvider);
    }

    public CreateInvoiceLink isFlexible(boolean isFlexible) {
        return add("is_flexible", isFlexible);
    }

}

File: library/src/main/java/com/pengrad/telegrambot/request/CreateInvoiceLink.kt
package com.pengrad.telegrambot.request

import com.pengrad.telegrambot.model.request.LabeledPrice
import com.pengrad.telegrambot.response.StringResponse
import com.pengrad.telegrambot.utility.kotlin.optionalRequestParameter
import com.pengrad.telegrambot.utility.kotlin.requestParameter

class CreateInvoiceLink(
    title: String,
    description: String,
    payload: String,
    currency: String,
    vararg prices: LabeledPrice
) : KBaseRequest<CreateInvoiceLink, StringResponse>(StringResponse::class) {

    /**
     * Backward compatibility: API 7.4, parameter "provider_token" became optional
     * @deprecated Use constructor without 'provider_token' instead
     */
    @Deprecated("Use constructor without 'providerToken' parameter", ReplaceWith("CreateInvoiceLink(title, description, payload, currency, *prices)"))
    constructor(
        title: String,
        description: String,
        payload: String,
        providerToken: String,
        currency: String,
        vararg prices: LabeledPrice
    ) : this(title, description, payload, currency, *prices) {
        this.providerToken = providerToken
    }

    val title: String by requestParameter(title)
    val description: String by requestParameter(description)
    val payload: String by requestParameter(payload)
    val currency: String by requestParameter(currency)
    val prices: Array<LabeledPrice> by requestParameter(arrayOf(*prices))

    var providerToken: String? by optionalRequestParameter()
    var subscriptionPeriod: Int? by optionalRequestParameter()
    var businessConnectionId: String? by optionalRequestParameter()
    var maxTipAmount: Int? by optionalRequestParameter()
    var suggestedTipAmounts: Array<Int>? by optionalRequestParameter()
    var providerData: String? by optionalRequestParameter()
    var photoUrl: String? by optionalRequestParameter()
    var photoSize: Int? by optionalRequestParameter()
    var photoWidth: Int? by optionalRequestParameter()
    var photoHeight: Int? by optionalRequestParameter()
    var needName: Boolean? by optionalRequestParameter()
    var needPhoneNumber: Boolean? by optionalRequestParameter()
    var needEmail: Boolean? by optionalRequestParameter()
    var needShippingAddress: Boolean? by optionalRequestParameter()
    var sendEmailToProvider: Boolean? by optionalRequestParameter()
    var sendPhoneNumberToProvider: Boolean? by optionalRequestParameter()
    var isFlexible: Boolean? by optionalRequestParameter()

    fun providerToken(providerToken: String) = apply { this.providerToken = providerToken }

    fun subscriptionPeriod(subscriptionPeriod: Int) = apply { this.subscriptionPeriod = subscriptionPeriod }

    fun businessConnectionId(businessConnectionId: String) = apply { this.businessConnectionId = businessConnectionId }

    fun maxTipAmount(maxTipAmount: Int) = apply { this.maxTipAmount = maxTipAmount }

    fun suggestedTipAmounts(suggestedTipAmounts: Array<Int>) = apply { this.suggestedTipAmounts = suggestedTipAmounts }

    fun providerData(providerData: String) = apply { this.providerData = providerData }

    fun photoUrl(photoUrl: String) = apply { this.photoUrl = photoUrl }

    fun photoSize(photoSize: Int) = apply { this.photoSize = photoSize }

    fun photoWidth(photoWidth: Int) = apply { this.photoWidth = photoWidth }

    fun photoHeight(photoHeight: Int) = apply { this.photoHeight = photoHeight }

    fun needName(needName: Boolean) = apply { this.needName = needName }

    fun needPhoneNumber(needPhoneNumber: Boolean) = apply { this.needPhoneNumber = needPhoneNumber }

    fun needEmail(needEmail: Boolean) = apply { this.needEmail = needEmail }

    fun needShippingAddress(needShippingAddress: Boolean) = apply { this.needShippingAddress = needShippingAddress }

    fun sendEmailToProvider(sendEmailToProvider: Boolean) = apply { this.sendEmailToProvider = sendEmailToProvider }

    fun sendPhoneNumberToProvider(sendPhoneNumberToProvider: Boolean) = apply { this.sendPhoneNumberToProvider = sendPhoneNumberToProvider }

    fun isFlexible(isFlexible: Boolean) = apply { this.isFlexible = isFlexible }

}